pkg/download/downloader.go

```diff
@@ -92,6 +107,38 @@ func extractZIP(targetDir string, read io.ReaderAt, size int64) error {
 	return nil
 }
 
+func isSymlinkAllowed(targetDir, oldname, newpath string) error {
+	// no symlinks to absolute paths
+	if filepath.IsAbs(oldname) {
+		glog.V(4).Infof("invalid absolute symlink %v -> %v\n", oldname, newpath)
+		return errors.New("invalid symlink referencing an absolute path in tar")
+	}
+
+	oldpath := filepath.Join(filepath.Dir(newpath), filepath.Clean(oldname))
+	fmt.Printf("LINK %s %s\n", oldpath, newpath)
+
+	if _, isNonEscaping := pathutil.IsSubPath(targetDir, oldpath); !isNonEscaping {
```

Comment:
This is going to be vulnerable to symlink chaining. E.g.

```
target = /a/b
tree:
/a/b/c
  c1 -> ../
  c2 -> c1/..
  c3 -> c2/..
```

c1 oldpath resolves to /a/b -> allowed
c2 oldpath resolves to /a/b/c -> allowed
c3 oldpath resolves to /a/b/c -> allowed

but c3 actually points to /a/b/c/../../.. = /

Reply:
thanks for the comments @tallclair! i think i understand risk, e.g.

```sh
echo "no sir" > /tmp/nope
mkdir -p /tmp/stage/inner && cd /tmp/stage/inner
ln -s .. l1 # <-- l1 points to /tmp/stage
ln -s l1/.. l2 # <-- l2 points to /tmp
cat l2/nope # <-- oops! because l2/nope is /tmp/nope
```

however, line 120 validates `oldpath`, *not* `oldname

in my two example links, oldname would be `..` and `l1/..`, respectively -- both seemingly safe

whereas oldpath would be `/tmp/stage` for the first symlink, which correctly passes the `IsSubPath` validator (i.e. `isNonEscaping==true`); whereas, for the second link, oldpath would be `/tmp` which would result `isNonEscaping==false`

i must be missing something, in my analysis, because i'm sure much deeper and more careful thought has been put into this problem than in my tea-infused morning thoughts.

Reply:
thanks for finding this. in go, `filepath.Clean` also behaves oddly, e.g. with

```sh
ln -s l1 ..
ln -s l1/.. l2
```

if we process the second symlink before the first, then `filepath.Clean("l1/..")` oddly returns ".". And so the currently logic in this PR validates l2, even if we want to disallow it as escaping. 

on the other hand, if we process these in creation order, then i think we can be safe (where we add a call to filepath.Abs, i.e. 
```go
oldpath, err := filepath.Abs(... what we currently assign to oldpath)
```

Reply:
I think @tallclair has meant this scenario:

```
c1 -> ../..
c2 -> ./c1/foo
```

in this case, c2 will be escaping the extract dir.

(It's expected filepath.Clean("foo/..") will return ".")

We need to make sure everything we pass to [IsSubPath](https://godoc.org/sigs.k8s.io/krew/pkg/pathutil#IsSubPath) is evaluated first if there are symlinks on the way (recursively).

Therefore in your code, we will need to `readlink` (which I am not sure if works recursively) on oldname as well as newpath.